File: app/src/main/overlays/history-overlay.ts
import type { OverlayManager, OverlayItem } from '../overlay-manager';
import type { CommandHistoryService } from '../services/command-history';
import type { ClaudeSession } from '../claude-session';

export function showHistoryOverlay(
  overlay: OverlayManager,
  history: CommandHistoryService,
  session: ClaudeSession,
  onUpdate: () => void,
): void {
  const entries = history.getRecent(30).reverse(); // newest first

  const items: OverlayItem[] = [];

  if (entries.length === 0) {
    items.push({ label: 'No commands recorded yet', dimmed: true });
  } else {
    for (const entry of entries) {
      const time = new Date(entry.ts);
      const hh = time.getHours().toString().padStart(2, '0');
      const mm = time.getMinutes().toString().padStart(2, '0');
      const cmd = entry.cmd;

      items.push({
        label: `${hh}:${mm}  ${cmd}`,
        action: () => {
          // Type the command into the terminal (don't auto-submit)
          if (session?.running) {
            session.write(cmd);
          }
          overlay.close();
        },
      });
    }
  }

  overlay.show({
    title: 'Command History',
    items,
    onClose: onUpdate,
    anchor: 'status-command',
    footer: 'Enter to type command into terminal',
  });
}

File: app-native/Sources/Orch/Overlays/HistoryOverlay.swift
import Foundation

func showHistoryOverlay(overlay: OverlayManager, history: CommandHistoryService, session: ClaudeSession, onUpdate: @escaping () -> Void) {
    let entries = history.getRecent(30).reversed()
    var items: [OverlayItem] = []

    if entries.isEmpty {
        items.append(OverlayItem(label: "No commands recorded yet", dimmed: true))
    } else {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"

        for entry in entries {
            let date = ISO8601DateFormatter().date(from: entry.ts) ?? Date()
            let time = formatter.string(from: date)
            items.append(OverlayItem(label: "\(time)  \(entry.cmd)", action: {
                if session.running { session.write(entry.cmd) }
                overlay.close()
            }))
        }
    }

    overlay.show(OverlayConfig(
        title: "Command History",
        items: items,
        onClose: onUpdate,
        footer: "Enter to type command into terminal",
        anchor: "status-command"
    ))
}
